diff --git a/mlx_lm/generate.py b/mlx_lm/generate.py
index e65016c..cace731 100644
--- a/mlx_lm/generate.py
+++ b/mlx_lm/generate.py
@@ -230,7 +230,7 @@ def wired_limit(model: nn.Module, streams: Optional[List[mx.Stream]] = None):
     async eval could be running pass in the streams to synchronize with prior
     to exiting the context manager.
     """
-    if not mx.metal.is_available():
+    if mx.default_device() != mx.gpu:
         try:
             yield
         finally:
